lib/concurrent/timer_task2.rb

```diff
@@ -0,0 +1,297 @@
+  #          puts 'Boom!'
+  #        end
+  #
+  #   task.execution_interval #=> 5
+  #   task.timeout_interval   #=> 5
+  #
+  # @example Immediate execution with `:run_now`
+  #   task = Concurrent::TimerTask2.new(run_now: true){ puts 'Boom!' }
+  #   task.execute
+  #
+  #   #=> 'Boom!'
+  #
+  # @example Controlling execution from within the block
+  #   channel = Concurrent::Promises::Channel.new 1
+  #   timer_task = Concurrent::TimerTask2.new(execution_interval: 1, channel: channel) do |task, cancellation|
+  #     task.execution_interval.times{ print 'Boom! ' }
+  #     print "\n"
+  #     task.execution_interval += 1
+  #     if task.execution_interval > 5
+  #       puts 'Stopping...'
+  #       task.shutdown
+  #     end
+  #   end
+  #
+  #   timer_task.execute # non-blocking call
+  #   6.times { channel.pop }
+  #   #=> Boom!
+  #   #=> Boom! Boom!
+  #   #=> Boom! Boom! Boom!
+  #   #=> Boom! Boom! Boom! Boom!
+  #   #=> Boom! Boom! Boom! Boom! Boom!
+  #   #=> Stopping...
+  #
+  # @example Observation
+  #   channel = Concurrent::Promises::Channel.new
+  #   # TODO: Make sure this works
+  #   observe = ->(channel) do
+  #    channel.pop_op.then do |(success, result, error)|
+  #       if result
+  #         print "(#{success}) Execution successfully returned #{result}\n"
+  #       elsif ex.is_a?(Concurrent::TimeoutError)
+  #         print "(#{success}) Execution timed out\n"
+  #       else
+  #         print "(#{success}) Execution failed with error #{ex}\n"
+  #       end
+  #     end
+  #   end
+  #
+  #   task = Concurrent::TimerTask2.new(execution_interval: 1, timeout_interval: 1, channel: channel){ 42 }
+  #   task.execute
+  #
+  #   3.times do
+  #     observe.call(channel)
+  #   end
+  #
+  #   #=> (true) Execution successfully returned 42
+  #   #=> (true) Execution successfully returned 42
+  #   #=> (true) Execution successfully returned 42
+  #   task.shutdown
+  #   channel.pop
+  #
+  #   task = Concurrent::TimerTask2.new(execution_interval: 1, timeout_interval: 1, channel: channel) do |_, cancellation|
+  #     raise Concurrent::TimeoutError.new if cancellation.cancelled?
+  #   end
+  #   task.execute
+  #
+  #   3.times do
+  #     observe.call(channel)
+  #   end
+  #   #=> (false) Execution timed out
+  #   #=> (false) Execution timed out
+  #   #=> (false) Execution timed out
+  #   task.shutdown
+  #   channel.pop
+  #
+  #   task = Concurrent::TimerTask2.new(execution_interval: 1, channel: channel){ raise StandardError }
+  #   task.execute
+  #
+  #   3.times do
+  #     observe.call(channel)
+  #   end
+  #   #=> (false) Execution failed with error StandardError
+  #   #=> (false) Execution failed with error StandardError
+  #   #=> (false) Execution failed with error StandardError
+  #   task.shutdown
+  #   channel.pop
+  #
+  # @see http://docs.oracle.com/javase/7/docs/api/java/util/TimerTask.html
+  class TimerTask2
```

Comment:
It should inherit from `Concurrent::Object` so it can use `safe_initialization!`

Reply:
Do you mean `::Concurrent::Synchronization::Object`?

Reply:
yeah, my bad